From e4fe96eb01f9a6d070189c1167f317f63781d521 Mon Sep 17 00:00:00 2001
From: Gerrit Pape <pape@smarden.org>
Date: Tue, 3 Jul 2007 09:11:39 +0000
Subject: [PATCH] gitk: properly resolve ambiguity if argument is both, revision and filename

If a repository contains a file and a branch with the same name, have
gitk recognize the '--' separator as final command line argument, as
documented in gitk(1); previously 'gitk <name> --' failed.  Additionally
have gitk insert '--' automatically for views created or edited through
the View menu.

The bug was noticed and reported by Josh Triplett through
 http://bugs.debian.org/425491

Signed-off-by: Gerrit Pape <pape@smarden.org>

Conflicts:

	gitk
---
 gitk |   15 +++++++++------
 1 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/gitk b/gitk
index 2d6a6ef..6329366 100755
--- a/gitk
+++ b/gitk
@@ -82,15 +82,12 @@ proc dorunq {} {
 proc start_rev_list {view} {
     global startmsecs
     global commfd leftover tclencoding datemode
-    global viewargs viewfiles commitidx
+    global viewargs viewfiles viewargsep commitidx
     global lookingforhead showlocalchanges
 
     set startmsecs [clock clicks -milliseconds]
     set commitidx($view) 0
-    set args $viewargs($view)
-    if {$viewfiles($view) ne {}} {
-	set args [concat $args "--" $viewfiles($view)]
-    }
+    set args [concat $viewargs($view) $viewargsep($view) $viewfiles($view)]
     set order "--topo-order"
     if {$datemode} {
 	set order "--date-order"
@@ -1659,7 +1656,7 @@ proc allviewmenus {n op args} {
 proc newviewok {top n} {
     global nextviewnum newviewperm newviewname newishighlight
     global viewname viewfiles viewperm selectedview curview
-    global viewargs newviewargs viewhlmenu
+    global viewargs viewargsep newviewargs viewhlmenu
 
     if {[catch {
 	set newargs [shellsplit $newviewargs($n)]
@@ -1683,6 +1680,7 @@ proc newviewok {top n} {
 	set viewperm($n) $newviewperm($n)
 	set viewfiles($n) $files
 	set viewargs($n) $newargs
+	set viewargsep($n) "--"
 	addviewmenu $n
 	if {!$newishighlight} {
 	    run showview $n
@@ -1702,6 +1700,7 @@ proc newviewok {top n} {
 	if {$files ne $viewfiles($n) || $newargs ne $viewargs($n)} {
 	    set viewfiles($n) $files
 	    set viewargs($n) $newargs
+	    set viewargsep($n) "--"
 	    if {$curview == $n} {
 		run updatecommits
 	    }
@@ -7437,10 +7436,12 @@ if {![file isdirectory $gitdir]} {
 }
 
 set cmdline_files {}
+set cmdline_sep ""
 set i [lsearch -exact $revtreeargs "--"]
 if {$i >= 0} {
     set cmdline_files [lrange $revtreeargs [expr {$i + 1}] end]
     set revtreeargs [lrange $revtreeargs 0 [expr {$i - 1}]]
+    set cmdline_sep "--"
 } elseif {$revtreeargs ne {}} {
     if {[catch {
 	set f [eval exec git rev-parse --no-revs --no-flags $revtreeargs]
@@ -7481,6 +7482,7 @@ set selectedhlview None
 set viewfiles(0) {}
 set viewperm(0) 0
 set viewargs(0) {}
+set viewargsep(0) {}
 
 set cmdlineok 0
 set stopped 0
@@ -7502,6 +7504,7 @@ if {$cmdline_files ne {} || $revtreeargs ne {}} {
     set viewname(1) "Command line"
     set viewfiles(1) $cmdline_files
     set viewargs(1) $revtreeargs
+    set viewargsep(1) $cmdline_sep
     set viewperm(1) 0
     addviewmenu 1
     .bar.view entryconf Edit* -state normal
-- 
1.5.2.2

